src/mutableAtom/index.ts

```diff
@@ -0,0 +1,138 @@
+import { atom } from 'jotai'
+import type { PrimitiveAtom } from 'jotai'
+import { proxy, snapshot, subscribe } from 'valtio'
+import type {
+  Action,
+  ActionWithPayload,
+  Options,
+  ProxyState,
+  Store,
+  Wrapped,
+} from './types'
+
+export function mutableAtom<Value>(
+  value: Value,
+  options: Options<Value> = defaultOptions
+) {
+  const valueAtom = atom({ value })
+
+  if (process.env.NODE_ENV !== 'production') {
+    valueAtom.debugPrivate = true
+  }
+  return makeMutableAtom(valueAtom, options)
+}
+
+export function makeMutableAtom<Value>(
+  valueAtom: PrimitiveAtom<Wrapped<Value>>,
+  options: Options<Value> = defaultOptions
+) {
+  const { proxyFn } = { ...defaultOptions, ...options }
+
+  const storeAtom = atom<
+    Store<Value>,
+    [ActionWithPayload<'setValueAsync', Value> | Action<'mount'>],
+    void
+  >(
+    (get, { setSelf }) => ({
+      proxyState: null,
+      getValue: () => get(valueAtom).value,
+      setValue: async (value: Value) => {
+        await defer()
+        setSelf({ type: 'setValueAsync', payload: value })
+      },
+    }),
+    (get, set, action) => {
+      if (action.type === 'setValueAsync') {
+        set(valueAtom, { value: action.payload })
+      } else if (action.type === 'mount') {
+        const store = get(storeAtom)
+        store.setValue = (value: Value) => {
+          set(valueAtom, { value })
+        }
```

Comment:
What happens if we remove this? It requires async update always, but that's the point. I think it should be consistent behavior, regardless of the mount status.

Reply:
I can think of a use-case where synchronous updates would be important.

```ts
it.only('should perform synchronous update', async () => {
  expect.assertions(4)
  const mutableCountAtom = mutableAtom(0)
  const countIsNotZeroAtom = atom((get) => get(mutableCountAtom).value > 0)
  const incrementCountAtom = atom(null, (get) => {
    const countProxy = get(mutableCountAtom)
    expect(get(countIsNotZeroAtom)).toBe(false)
    countProxy.value++
    expect(get(countIsNotZeroAtom)).toBe(true)
  })
  let isMounted = false
  incrementCountAtom.onMount = () => {
    isMounted = true
  }
  function useTest() {
    useAtom(countIsNotZeroAtom)
    const [, incrementCount] = useAtom(incrementCountAtom)
    return { incrementCount }
  }
  const { result } = renderHook(useTest)
  await waitFor(() => assert(isMounted))
  await act(async () => {
    result.current.incrementCount()
  })
})
```

Reply:
Since regular updates are synchronous, it makes sense to me that proxy updates should also be synchronous.

The only reason I added async updates was to support the edge case where proxy might be updated via a useEffect on component mount. However my base assumption that useEffect would fire before atom mount was incorrect. So it would possible to remove the async subscribe logic

Reply:
Oops, I misguided something and confused myself. Yes, I think everything should be sync and we don't need onMount.